Posix implementation layer

youvedeep opened the pull request:
Currently Zephyr does not support POSIX APIs. This patch patch provides a compatibility layer for supporting  POSIX1003.1 PSE52.  patch series use Zephyr APIs as foundation and provides a layer to map POSIX APIs to ZEPHYR APIs.

youvedeep-singh committed:
kernel: POSIX: Compatibility layer for pthread APIs.

This patch provides pthread APIs for POSIX 1003.1 PSE52 standard.

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

youvedeep-singh committed:
kernel: POSIX: moving POSIX related typedef into sys/types.h file.

As per POSIX standard typedef should be part of sys/types.h file.
So moving typedef from pthread.h to sys/types.h file.

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

youvedeep-singh committed:
kernel: POSIX: Compatibility layer for scheduler APIs.

This patch provides scheduler APIs for POSIX 1003.1 PSE52 standard.

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

youvedeep-singh committed:
kernel: POSIX: Compatibility layer for POSIX sleep APIs.

This patch provides POSIX sleep APIs for POSIX 1003.1 PSE52 standard.
sleep(n) is implemented using Zephyr k_sleep API.
uleep(n) is implemented using Zephyr k_sleep/k_busy_Wait API.

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

aescolar-ot commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,138 @@
+
+}
+
+int pthread_attr_setschedparam(pthread_attr_t *attr,
```
Please add all these functions also to posix_cheats.h
Otherwise you prevent the POSIX arch port from using the host ones.

aescolar-ot commented on kernel/posix/pthread_sched.c:
```diff
@@ -0,0 +1,59 @@
+#include <kernel.h>
+#include <sched.h>
+
+int sched_get_priority_min(int policy)
```
Also these 2, and any other which is the POSIX API

lpereira commented on include/posix/pthread.h:
```diff
@@ -17,29 +18,37 @@ struct timespec {
+ *  Default pthread attributes on attributes initialization.
+ */
+static pthread_attr_t initpthreadattrs = {
+	K_LOWEST_APPLICATION_THREAD_PRIO,	/* priority */
```
Please see comments on #5494 about this.

lpereira commented on kernel/Kconfig:
```diff
@@ -500,9 +500,19 @@ config PTHREAD_IPC
+if PTHREAD_IPC
+config MAX_PTHREAD_COUNT
+	int
+	prompt "Maximu pthread count in POSIX application"
```
Typo: maximu

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,138 @@
+int pthread_create(pthread_t *newthread, const pthread_attr_t *attr,
+		   void *(*threadroutine)(void *), void *arg)
+{
+	static int thread_count;
```
This should be an atomic_val_t.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,138 @@
+		errno = EINVAL;
+		return -1;
+	}
+	if (thread_count < CONFIG_MAX_PTHREAD_COUNT) {
```
Between creating a thread and incrementing the value, preemption might occur, causing a race condition that will lead to invalid memory access.

`atomic_inc()` will return the previous value, which is the one you want to use:

```C

atomic_val_t index = atomic_inc(&thread_count);
if (index < CONFIG_MAX_PTHREAD_COUNT) {
    ...  &threads[index] ...
}
```

This way, `index` is guaranteed to be below `CONFIG_MAX_PTHREAD_COUNT`.  This way, there's no race condition: there's no way to have an element in the threads array that's going to be skipped, or two threads that are initialized in the same element.

youvedeep replied:
Thanks @lpereira I will take care of this.

lpereira reviewed (changes requested):
Please review my comments on the other pthread PR; some of the issues there were not addressed in this PR.

(I did not review this fully.)

youvedeep-singh committed:
kernel: POSIX: Compatibility layer for POSIX clock APIs.

This patch provides POSIX clock APIs for POSIX 1003.1 PSE52 standard.

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

youvedeep commented:
Addressed comments from @lpereira  and @aescolar-ot in new patch version.

lpereira commented on kernel/posix/pthread_sched.c:
```diff
@@ -0,0 +1,58 @@
+
+int sched_get_priority_min(int policy)
+{
+	if (!((policy == SCHED_FIFO) || (policy == SCHED_RR))) {
```
You can apply [De Morgan's Law](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) here and simplify this expression.

youvedeep replied:
Sure I will change it to  - if((policy != SCHED_FIFO) && (policy != SCHED_RR)))

lpereira replied:
No need for the extra parenthesis.

lpereira commented on kernel/posix/pthread_sched.c:
```diff
@@ -0,0 +1,58 @@
+
+	if (IS_ENABLED(CONFIG_COOP_ENABLED)) {
+		if (policy == SCHED_FIFO) {
+			errno = 0;	/* Valid negative return */
```
This function can't return a negative value on success; that's how the API is defined, and how code using it will expect it to behave.

A solution is not to have priorities mapping 1:1 between Zephyr and POSIX.  You can derive the formula for the proper mapping like so:

- Consider two points in the space: (x0, y0) and (x1, y1).  The X axis is the POSIX priority, and the Y axis is the Zephyr priority.
- Find two points, P1 and P2, that maps the minimum priority in Zephyr (that's a negative value) to a value that's acceptable in POSIX (say, 1). Do the same for P2, for the maximum value.
- Calculate the slope: m = (y1 - y0) / (x1 - x0)
- Use the calculated slope in the line equation: y - y0 = m(x - x0), where y0 and x0 are the coordinates from P1
- Isolate y or x depending on which direction you want to go (Zephyr -> POSIX or POSIX -> Zephyr).  If you're calculating y, any x in the line is valid (could be even x0), and vice versa.
- It's important to round m to the nearest integer, to not use floating point math here.  It shouldn't matter if it's ceiled or floored

youvedeep replied:
As per my understanding if a function returns a negative value then it should set errno to 0.
because of which I was putting 0 as ERROR code.

 https://www.unix.com/man-page/posix/3posix/sched_get_priority_min/

lpereira replied:
That's the opposite: if the function returns -1, errno must be set; otherwise, errno isn't touched (see **RETURN VALUE** in the linked man page).

Per the POSIX specification, only a return value of -1 will signal error.  However, a lot of code will verify error conditions by testing the sign bit (i.e. checking if the return value is < 0 rather than == -1), as it often generates slightly tighter code.

Being a compatibility layer, it's best to assume that any function returning -1 per the specification will also return a negative number (whatever it is) on error, and 0 or a positive value on success.  Returning any value from 0 to INT_MAX is fine for these functions, as long as they're consistent with each other, and that the functions to set the priorities are also consistent with these priority values.

youvedeep replied:
Looking at documentation @ https://www.systutorials.com/docs/linux/man/2-sched_get_priority_max/#index, (Looks like man page for IEEE Std 1003.13-2003 is silent on number of priorities level  in POSIX https://www.unix.com/man-page/posix/3posix/sched_get_priority_min/)- 
As per POSIX :- 
1. Higher the numerical value Priority, Higher the priority. Something like a thread with priority 5 will have higher priority than a thread with priority 1. (something opposite of Zephyr).
 2. There should be at least 32 priority levels, but this can vary based on other standards. I think we can avoid this as for RTOS this may be too high number of priority levels.

I think we can not use the slope technique to map Zephyr priort levels to POSIX prioirty levels as this may create confusion/issues.  Let us assume (Numerical lower priority has Higher Priority than numberical high only for example). 
Let us take an example - If we have premptive based scheduling enabled with 
Zephyr_Max_ priority = 15
Zephyr_Min_ priority = 0
Now mapping this range to POSIX_MAX = 31, POSIX_MIN=0.
Let us say that a POSIX compliant application has 3 threads with priority 2, 3, 4. then it may be expected that   1 has higher priority than 2 and so on.
But internally :- 
Thread 1 ==> POSIX_PPRIOITY = 2 mapped to ZEPHYR_PRIORTY = 1
Thread 2 ==> POSIX_PPRIOITY = 3 mapped to ZEPHYR_PRIORTY = 1
Thread 3 ==> POSIX_PPRIOITY = 4 mapped to ZEPHYR_PRIORTY = 2
So this may give undesirable results.  So i think there should be 1:1 or n:m (n>= m) mapping between POSIX and ZEPHYR .

I think we can map ZEPHYR_PRIORITY to POSIX_PRIORTY :-
For Preemptive_thread :-  
posix_priorty= (CONFIG_NUM_PREEMPT_PRIORITIES - 1) - zephyr_priorty
*-1 to keep maximum priority reserved for highest priority thread.

This will make POSIX_MIN_PRIORITY = 0 and POSIX_MAX_PRIORTY = CONFIG_MAX_APPLICATION_PRIORITY
For COOPERATIVE scheduling  policy:- 
if preemption Enabled :- 
posix_application_priority = CONFIG_NUM_COOP_PRIORITIES  + zephyr_priority. (cooperative priorities are -ive).
if preemption Enabled :- 
posix_application_priority = (CONFIG_NUM_COOP_PRIORITIES - 1)  + zephyr_priority. (cooperative priorities are -ive).
*-1 to keep maximum priority reserved for highest priority thread.

lpereira commented on include/posix/pthread.h:
```diff
@@ -359,4 +381,117 @@ int pthread_barrierattr_getpshared(const pthread_barrierattr_t *, int *);
+ *
+ * See IEEE 1003.1
+ */
+static inline void pthread_exit(void *retval)
```
Do we have a `pthread_join()`?

youvedeep replied:
As of now we do not have pthread_join(). We need to figure out how to implement it. Before that we will target other POSIX APIs.

lpereira replied:
One way to do this is the following:

```C
struct pthread {
    struct k_thread thread;
    void *retval;
};

typedef pthread_t struct pthread;
```

This way, it's possible to use type-punning to convert between a `struct k_thread` and a `struct pthread`.  And, for pthreads, it's possible to set/obtain the return value.  So, this function could be implemented like so:

```C
static void pthread_exit(void *retval)
{
        pthread_t *self = (pthread_t *)k_current_get();

        self->retval = retval;

        k_thread_abort(self);
}
```

Then, `pthread_join()` will get the `retval` from `pthread_t`.  Calling `pthread_exit()` should also give a mutex that `pthread_join()` is waiting on, before it aborts itself.

youvedeep replied:
Yes I got your Idea, but I wanted to enhance it to be compatible with pthread_join after first set of patch merge. 

For making pthread_join() we need to have 2 members (mutex and Condition variable to signal the state exit). But before that i am waiting for my first set of patches to be merged. I am planning to take this as enhancement, 
struct pthread {
    struct k_thread thread;
    void *retval;
     pthread_mutex_t state_lock;   /* Locks the state.  */
     pthread_cond_t state_cond;    /* Signalled when the state changes.  */
};

lpereira commented:
There are some spelling/typos in the commit messages as well; please fix them.

andrewboie commented on include/posix/pthread.h:
```diff
@@ -7,39 +7,37 @@
+#define PTHREAD_CANCELABLE     (0 << PTHREAD_CANCEL_POS)
+/* Essential Thread */
+#define PTHREAD_NON_CANCELABLE (1 << PTHREAD_CANCEL_POS)
+#define K_LOWEST_STACK_SIZE     256
```
how was this value arrived at?

youvedeep replied:
I just went through the Zephyr Code base and looked that 256 is least stack size used by a thread.

youvedeep replied:
Hi @andrewboie Sorry for confusion. But i have mentioned clearly on PR https://github.com/zephyrproject-rtos/zephyr/pull/5494 that discussion for this patch is moved to https://github.com/zephyrproject-rtos/zephyr/pull/5667.
@nashif requested to merge all POSIX related patches into 1 PR and for that I created this PR later.
To avoid confusion, I have closed other POSIX related PRs.

andrewboie replied:
You don't need this value at all unless I'm missing something.
It only seems to be used for the defaults which have a NULL stack pointer anyway,
Just use NULL for the default stack pointer and 0 for the size. There is no actual minimum. I have a stack in a double fault handler which is about 12 bytes in size.

youvedeep replied:
hmm got it, So basically your suggestion to to define stack_size as 0, something like stack pointer. Let the user provide stack_size along with stack pointer,

andrewboie replied:
> Hi @andrewboie Sorry for confusion. But i have mentioned clearly on PR #5494 that discussion for this patch is moved to #5667.

Don't have multiple open PRs with the same code. Just don't do it. Close one of them.

tyrantlizard77 reviewed (changes requested):
Why is this PR and #5494 both open for review at the same time, they have the same code in them!
This is wasting developer time, clean up your stuff and send for review in one or more PRs that are completely disjoint.

youvedeep commented:
Guys, Can you please please provide quick review comments on this PR as this is targeted for Release 1.11

lpereira commented on include/posix/pthread.h:
```diff
@@ -7,39 +7,42 @@
+/* Pthread cancellation */
+#define PTHREAD_CANCEL_POS     0
+/* Non Essential Thread */
+#define PTHREAD_CANCELABLE     (0 << PTHREAD_CANCEL_POS)
```
I can't seem to find any reference to `PTHREAD_CANCELABLE`. Is this documented somewhere?

lpereira commented on include/posix/pthread.h:
```diff
@@ -7,39 +7,42 @@
+#define PTHREAD_CANCELABLE     (0 << PTHREAD_CANCEL_POS)
+/* Essential Thread */
+#define PTHREAD_NON_CANCELABLE (1 << PTHREAD_CANCEL_POS)
+#define THREAD_INIT_FLAGS       PTHREAD_CANCELABLE
```
Why is this prefixed with `THREAD_`?  If someone includes `pthread.h`, this will pollute their namespace with things that are not POSIX-specific. There doesn't seem to exist any `THREAD_INIT_FLAGS` in Linux as well, so I don't know why is this defined in the header file.

lpereira commented on include/posix/pthread.h:
```diff
@@ -7,39 +7,42 @@
-struct timespec {
-	s32_t tv_sec;
-	s32_t tv_nsec;
+static const pthread_attr_t initpthreadattrs = {
```
Nitpick, but please use `_` to separate words in variable names: `init_pthread_attrs`.

Also, this shouldn't be in the header file; there will be multiple copies of it if it's included multiple times.  Define it inside `pthread_attr_init()` instead, even if you have to move it to a `.c` file.

lpereira commented on include/posix/pthread.h:
```diff
@@ -383,4 +376,163 @@ int pthread_barrierattr_getpshared(const pthread_barrierattr_t *, int *);
+	k_thread_abort((k_tid_t) self);
+}
+
+static inline u32_t zephyr_to_posix_priority(s32_t z_prio, int *policy)
```
Functions that are implementation details should be in the `.c` file.

lpereira commented on include/posix/time.h:
```diff
@@ -0,0 +1,44 @@
+ */
+static inline int clock_settime(clockid_t clock_id, const struct timespec *ts)
+{
+	errno = (clock_id != CLOCK_MONOTONIC) ? EINVAL : ENOSYS;
```
The error here should always be `ENOSYS`, since it's just a stub. Returning `EINVAL` does not help the user figure out what's wrong.

lpereira commented on include/posix/time.h:
```diff
@@ -0,0 +1,44 @@
+int clock_gettime(clockid_t clock_id, struct timespec *ts);
+
+/**
+ * @brief POSIX threading compatibility API
```
This `@brief` looks like it was copy-pasted from other pthread-related documentation.

lpereira commented on include/posix/time.h:
```diff
@@ -0,0 +1,44 @@
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
```
Might be better to just point out to a [online copy of a man page](http://pubs.opengroup.org/onlinepubs/7908799/xsh/clock_settime.html) for `clock_settime()` instead.

lpereira replied:
If linking to the man page, please don't forget to document the differences in Zephyr. In particular, for this API, document the clock_id type that is supported.

lpereira commented on kernel/posix/clock.c:
```diff
@@ -0,0 +1,29 @@
+#include <time.h>
+#include <errno.h>
+
+int clock_gettime(clockid_t clockId, struct timespec *ts)
```
noCamelCaseInZephyr

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,159 @@
+		   void *(*threadroutine)(void *), void *arg)
+{
+	static atomic_val_t thread_count;
+	int thread_index = 0;
```
Use an `atomic_val_t` here: that's what `atomic_inc()` returns.  Even though it's just a typedef to int, it's better to always use the correct type to future-proof the code, readability, and avoid gotchas with implicit integer conversions/losing precision.

lpereira commented on include/posix/pthread.h:
```diff
@@ -7,39 +7,42 @@
+/* Non Essential Thread */
+#define PTHREAD_CANCELABLE     (0 << PTHREAD_CANCEL_POS)
+/* Essential Thread */
+#define PTHREAD_NON_CANCELABLE (1 << PTHREAD_CANCEL_POS)
```
If I search for `PTHREAD_NON_CANCELABLE`, I only find a reference to this pull request. If this is an implementation detail, please move its definition to the `.c` file.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,159 @@
+	k_tid_t *thread = (k_tid_t *)pthread;
+	int new_prio, cur_prio;
+
+	if ((policy != SCHED_RR) && (policy != SCHED_FIFO)) {
```
Nitpick: please remove the excessive parenthesis.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,159 @@
+	new_prio = posix_to_zephyr_priority(param->priority, policy);
+	cur_prio = k_thread_priority_get(*thread);
+
+	if ((is_posix_prio_valid(new_prio, policy) == false) ||
```
Nitpick: please remove the excessive parenthesis.

lpereira commented on kernel/posix/pthread_sched.c:
```diff
@@ -0,0 +1,66 @@
+
+static bool valid_posix_policy(int policy)
+{
+	if ((policy != SCHED_FIFO) && (policy != SCHED_RR)) {
```
Nitpick: please remove the excessive parenthesis.

youvedeep replied:
I feel removing internal braces (if (policy != SCHED_FIFO && policy != SCHED_RR) ) will lose the readability. So keeping the braces is better. What is your thought ?

lpereira replied:
I don't have strong opinions on this -- I find it less visually polluted without the parenthesis, and very readable.  It's just a nitpick, we can agree to disagree.

lpereira commented on kernel/posix/pthread_sched.c:
```diff
@@ -0,0 +1,66 @@
+
+	return true;
+}
+int sched_get_priority_min(int policy)
```
Nitpick: Missing newline before function.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,159 @@
+	prio = posix_to_zephyr_priority(attr->priority, attr->schedpolicy);
+
+	thread_index = atomic_inc(&thread_count);
+	if (thread_index < CONFIG_MAX_PTHREAD_COUNT) {
```
BTW, I was thinking: can't we just use a memory pool for this? It's possible to return a handle to a pool when the thread ends. Doing it this way, there's no way to recycle a pthread struct: once the system runs out of them, they're gone.

youvedeep replied:
Hi @lpereira I will take this as enhancement after M2 milestone.

lpereira commented on tests/kernel/pthread/src/pthread.c:
```diff
@@ -181,7 +172,34 @@ int barrier_test_done(void)
+	}
+
+	for (i = 0; i < N_THR; i++) {
+		/* Thread Attribute initialization */
```
Please avoid comments that are not adding any value. If a line says `foo_init(&foo)`, don't say `/* Initializes foo */`. Comments should document the intent of the code or things that are not obvious just from reading the code itself.

lpereira commented:
The patches are looking better, but I don't think they're ready to be merged yet.

youvedeep commented:
Addressed comments from @laperie  in new patch set.

youvedeep-singh committed:
tests: kernel: posix: pthread_join: Add pthread join test.

Implement pthread_join test application.

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

youvedeep-singh committed:
tests: kernel: posix: clock: Add posix clock test.

Add posix clock API test.

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

youvedeep commented:
Addressed All comments from @lpereira  (except 1 for mslab for thread creation).

Can you please provide quick feedback if any, as we need this in V1.11 .

youvedeep commented:
Added some more test APIs into pthread_join patch (tests: posix: Add pthread join test.)

lpereira commented on include/posix/pthread.h:
```diff
@@ -7,38 +7,58 @@
+	int cancel_state;
+	int cancel_pending;
+	struct k_sem cancel_lock_sem;
+	pthread_mutex_t cancel_lock;
```
Why do we have a semaphore and a pthread_mutex?

youvedeep replied:
pthread_mutex_t just contains a pointer to a semphore. 
So we need a semaphore to for this.

lpereira replied:
A user using these APIs will have to define a k_sem as well?

youvedeep replied:
pthread mutex was implemented by @andyross in last zephyr release.

As per my understanding, there are 2 ways to achieve this :- 
1. Application need to define it statically something like :  "PTHREAD_MUTEX_DEFINE(lock);" (something shown in same application tests/kernel/pthread)
2. Application need to define a "k_sem" and then use ptherad_mutex_init to initialize it. (approach I have taken for my implementation for achieving cancel_lock and state_lock).

I think we need to modify this later, as it does not seem intuitive to a app developer.

lpereira replied:
Can we make it work for both cases (by statically initializing it with that macro or calling pthread_init()) if pthread_mutex_t were embedding a k_sem rather than pointing to it?

andrewboie replied:
> I think we need to modify this later

I think we need to fix this now.

youvedeep replied:
Yes @andrewboie we need to fix this now. But it will need changes in almost all mutex related APIs. At this stage I am avoiding this because merge window is approaching and this will gate all POSIX patches. I will put a FIXME: Comment here and will take this up after 1.11.

lpereira commented on include/posix/pthread.h:
```diff
@@ -383,4 +393,151 @@ int pthread_barrierattr_getpshared(const pthread_barrierattr_t *, int *);
+/* Base Pthread related APIs */
+
+/**
+ * @brief get pid of current execueting thread.
```
pthread_self() is not necessarily a PID; in fact, Zephyr doesn't have the notion of process ID.  `man 3p pthread_self` here says "get the calling thread ID", which I think is more appropriate.

youvedeep replied:
Addressed

lpereira commented on include/posix/pthread.h:
```diff
@@ -383,4 +393,151 @@ int pthread_barrierattr_getpshared(const pthread_barrierattr_t *, int *);
+}
+
+/**
+ * @brief POSIX pthread attributes destroy
```
`Destroys thread atrtibute object`

youvedeep replied:
Addressed

lpereira commented on include/posix/pthread.h:
```diff
@@ -383,4 +393,151 @@ int pthread_barrierattr_getpshared(const pthread_barrierattr_t *, int *);
+}
+
+/**
+ * @brief POSIX get pthread scheduling parameter
```
Do we need to prefix all `@brief` for these things with POSIX, and still mention `pthread`? Might be better to just say `Returns the schedparam attribute from a thread attribute object`, or something along these lines.  This applies to other functions here.

youvedeep replied:
Addressed.

lpereira commented on include/posix/pthread.h:
```diff
@@ -383,4 +393,151 @@ int pthread_barrierattr_getpshared(const pthread_barrierattr_t *, int *);
+	return 0;
+}
+/**
+ * @brief are Both pthreads same
```
`Compare thread IDs`.  Although the punctuation here is missing (questions need to end in a question mark), this must be a statement, not a question.  Please be consistent with the case as well (only proper nouns or sentences should have their first character uppercase, and acronyms should be written in all uppercase).

youvedeep replied:
Addressed.

lpereira commented on include/posix/pthread.h:
```diff
@@ -383,4 +393,151 @@ int pthread_barrierattr_getpshared(const pthread_barrierattr_t *, int *);
+}
+
+/**
+ * @brief get pid of current execueting thread.
```
Typo: `execueting`

Also, this is a function to set the schedpolicy for a Pthread attribute object.

youvedeep replied:
Addressed

lpereira commented on include/posix/pthread.h:
```diff
@@ -383,4 +393,151 @@ int pthread_barrierattr_getpshared(const pthread_barrierattr_t *, int *);
+}
+
+/**
+ * @brief get pid of current execueting thread.
```
Same comment for the set function.

youvedeep replied:
Addressed

lpereira commented on include/posix/pthread.h:
```diff
@@ -383,4 +393,151 @@ int pthread_barrierattr_getpshared(const pthread_barrierattr_t *, int *);
+static inline int pthread_attr_setdetachstate(pthread_attr_t *attr,
+					       int detachstate)
+{
+	if ((attr == NULL) || (detachstate != PTHREAD_CREATE_DETACHED &&
```
The handling of attr being a null pointer is irregular here.  Either check everywhere (could be an assert), or don't check at all: the standard says that the behavior is undefined if attr does not refer to an initialized thread attributes object.

You might want to study the [nonnull attribute](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#Common-Function-Attributes) in GCC as an alternative.

youvedeep replied:
Addressed

lpereira commented on include/posix/pthread.h:
```diff
@@ -383,4 +393,151 @@ int pthread_barrierattr_getpshared(const pthread_barrierattr_t *, int *);
+ *
+ * See IEEE 1003.1
+ */
+static inline int pthread_attr_getdetachstate(const pthread_attr_t *attr,
```
Is there any reason for these getter/setters to be inline?  They're very likely to be used rarely (usually once during system bootup), and are very likely not a bottleneck.

youvedeep replied:
Addressed.

lpereira commented on include/posix/unistd.h:
```diff
@@ -0,0 +1,46 @@
+ */
+static inline int usleep(useconds_t useconds)
+{
+	/* Creating Busy wait for faction of msecs */
```
Typo: faction

lpereira commented on include/posix/unistd.h:
```diff
@@ -0,0 +1,46 @@
+#include "sys/types.h"
+
+/**
+ * @brief suspend execuetion for specified seconds.
```
Typo: execuetion

I'm not a native speaker, but shouldn't this be `suspend execution of current thread for the specified number of seconds`?

lpereira commented on kernel/posix/clock.c:
```diff
@@ -0,0 +1,32 @@
+	elapsed_msecs = k_uptime_get();
+	elapsed_secs = elapsed_msecs / MSEC_PER_SEC;
+
+	/* For better accuracy for time in ns */
```
I'm not sure what this comment is telling me.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,351 @@
+#include "wait_q.h"
+
+/* Pthread structure for number of threads */
+struct posix_thread pthread[CONFIG_MAX_PTHREAD_COUNT];
```
This must be `static`.

youvedeep replied:
Addressed.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,351 @@
+	thread_index = atomic_inc(&thread_count);
+
+	if (thread_index < CONFIG_MAX_PTHREAD_COUNT) {
+		pthread[thread_index].cancel_state = (1 << PTHREAD_CANCEL_POS) &
```
It's cleaner to just get a pointer to `pthread[thread_index]` and set `thread->cancel_state`, `thread->state`, `thread->cancel_pending`, etc.

Might want to add a `FIXME` comment saying that this will change to a mempool (although it's, IMO, faster to just write it as a mempool right away; it's simple and will fix the issue of running out of `pthread` handles to give out if threads are created and destroyed).

youvedeep replied:
Addressed.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,351 @@
+	*oldstate = pthread->cancel_state;
+	pthread->cancel_state = state;
+
+	key = irq_lock();
```
I reckon that some of the pthread stuff will eventually be system calls. User mode code can't lock interrupts.

youvedeep replied:
Addressed. I removed irq_lock and using cancel_state mutex for this.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,351 @@
+		return -1;
+	}
+
+	memcpy((void *)attr, (const void *)&init_pthread_attrs,
```
Are the casts necessary? `attr` is not const, and any pointer can be converted to `void`. `init_pthread_attrs` is const already, so it should be possible to pass it to the second memcpy() parameter. While the memcpy() is fine, you can write this as `*attr = init_pthread_attrs;`, and get rid of the casts.

youvedeep replied:
Yes I was using "*attr = init_pthread_attrs;" instead of memcpy earlier but I get feedback that memcpy will be better choice as compare to direct assignment.

lpereira commented on include/posix/sys/types.h:
```diff
@@ -0,0 +1,73 @@
+	int priority;
+	void *stack;
+	size_t stacksize;
+	u32_t options;      /* Thread related Flags */
```
Why not just name `options` as something like `thread_flags`?

lpereira reviewed (changes requested):
This iteration is much better, thanks -- I still need to do a once over, but there are still a few things that need to be addressed

youvedeep commented on include/posix/pthread.h:
```diff
@@ -7,38 +7,58 @@
+#include "posix_sched.h"
+#include "unistd.h"
+
+/* Thread state. */
```
Addressed

youvedeep commented on include/posix/posix_sched.h:
```diff
@@ -0,0 +1,45 @@
+
+/* Priority based prempetive scheduling policy */
+#ifndef SCHED_RR
+#define SCHED_RR 1
```
Currently  SCHED_RR and SCHED_FIFO are not used in our Zephyr Code so there is not possibility of getting these assigned to different value. Also in Linux too it is directly defined.
I wanted to avoid the complicated logic of multiple conditions here.

youvedeep commented:
Addressed comments from @lpereira in latest patch set.

lpereira commented:
There are some spelling mistakes in comments and commit messages. Please comb them.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,550 @@
+	int priority = schedparam->priority;
+
+	if (!attr || !attr->initialized) {
+		errno = EINVAL;
```
pthread functions do not set errno, they return the error code directly.

youvedeep replied:
initially I was thinking of return -1 and send errno on Error, but looks like to be strictly compliant with POSIX, i will return -ERROR_CODE directly. I will modify this for all functions.

lpereira replied:
It's ERROR_CODE; not -ERROR_CODE. Yes, I know it's confusing and inconsistent with other APIs in POSIX, but that's how it's defined.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,550 @@
+	pthread_condattr_t cond_attr;
+	struct posix_thread *thread;
+
+	if (!attr || !attr->initialized || !attr->stack || !attr->stacksize) {
```
attr can be NULL; if that's the case, the thread is created with default attributes.

youvedeep replied:
Yes POSIX allow NULL attribute for new thread creation, but we will have issue with STACK allocation. As we do not have dynamic stack creation in Zephyr so stack should be passed wihle creating the thread.  Sure I will document in comment.

youvedeep replied:
Sure I will file a bug on this and will mention as a comment too in this.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,550 @@
+	void * (*fun_ptr)(void *);
+
+	fun_ptr = arg3;
+	(*fun_ptr)((void *)arg1);
```
Shouldn't the pthread state be set to terminated after fun_ptr() returns?

Also, nitpick, but this syntax of calling function pointers isn't used throughout Zephyr; prefer the shorthand syntax (`fun_ptr(arg1)`; no cast needed for arg1, it's already `void*`).

youvedeep replied:
Sure I will change it to fun_ptr(arg1), it is more readable too.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,550 @@
+	pthread->cancel_state = state;
+
+	if (state == PTHREAD_CANCEL_DISABLE) {
+		_thread_essential_set();
```
While using the cancel_lock mutex is fine, saying that a thread is essential is only possible in supervisor mode. While this is fine, it's one of the things that must be documented.

youvedeep replied:
So what is your suggestion, Shall we stop propagating the THREAD_CANCEL_ENABLE/DISABLE calls to zephyr and handle it in POSIX layer only ? 
Also when you mentioned documentation - Shall we document inline to code only or you suggest to create documentation separately and put it under "doc" directory in Zephyr.

lpereira replied:
It's fine to keep it this way. Only document that this function won't work in user mode. We can revisit this at a later date.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,550 @@
+			pthread_mutex_unlock(&thread->cancel_lock);
+			k_thread_abort((k_tid_t) thread);
+		} else {
+			thread->retval = PTHREAD_CANCELED;
```
Shouldn't this be thread->state?

youvedeep replied:
Yes I updated this - 
If a thread with CANCEL_ENABLE and JOINABLE is cancelled, It broadcast the join condition (for threads waiting to join it) and change the state to PTHREAD_EXIT. (something like thread_exit()), Please give feedback if this is ok we need to make some changes in this ?

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,550 @@
+		pthread_mutex_unlock(&self->state_lock);
+	} else {
+		pthread_mutex_unlock(&self->state_lock);
+		self->state = PTHREAD_TERMINATED;
```
Shouldn't the state be set before the mutex is unlocked?

youvedeep replied:
Thanks for catching this.

Yes we should change the state only when it is protected through state_lock.
I will make this change.

tyrantlizard77 reviewed (commented):
Worth noting as this came up in the recent memory protection discussions: although not a pre-condition for this patch, we actually do have use-cases for a k_thread_join() system call, which waits for the target thread object to terminate.

So if you wanted to implement k_thread_join() and then have pthread_join() be a thin wrapper on top of it that would be pretty cool.

You don't have to do this though. I think @agross-linaro is going to look into it eventually when he has time.

jmtorrespalma commented on include/posix/sys/types.h:
```diff
@@ -0,0 +1,55 @@
+
+#ifdef CONFIG_PTHREAD_IPC
+/* Mutex */
+typedef struct pthread_mutex {
```
Can someone explain why this is not simply `typedef k_sem pthread_mutex_t;`?
Current implementation requires a k_sem instance to be created too, while it also pollutes application namespace defining `struct pthread_mutex`.
Same applies for other typedefs.

andrewboie commented on arch/posix/include/posix_cheats.h:
```diff
@@ -30,7 +30,11 @@
 #define pthread_condattr_t     zap_pthread_condattr_t
 #define pthread_barrier_t      zap_pthread_barrier_t
 #define pthread_barrierattr_t  zap_pthread_barrierattr_t
+#define pthread_attr_t         zap_pthread_attr_t
```
why is this #define instead of a typedef?
why do we have this zap_ layer of abstraction?

youvedeep replied:
Hi  @andrewboie, 

Sorry for my limited knowledge on this. I just followed the last followed the earlier one like pthread_barrierattr_t. It was suggested by @aescolar-ot to put it this way.

_zap abstraction layer was put to avoid the conflict of POSIX layer calls with native_posix architecture.

aescolar-ot replied:
Just a comment: This posix_cheats.h header, and its defines, are meant to rename any Zephyr global symbol (function or variable) whose name conflicts with the native/host POSIX API, so that they will have a different name when compiling for ARCH_POSIX.

Regarding types, we just need to ensure the headers, which include types conflicting with the POSIX API types, are not included by mistake from anything in arch/posix/ or boards/posix/ (or any of its drivers which may rely on talking with the real host APIs). So, we should not have these types included in any basic kernel header which ends up included almost everywhere else.

andrewboie commented on include/posix/sys/types.h:
```diff
@@ -0,0 +1,74 @@
+/* Mutex */
+typedef struct pthread_mutex {
+	struct k_sem *sem;
+} pthread_mutex_t;
```
why is this not  "typedef struct k_sem pthread_mutex_t"

youvedeep replied:
@andyross  is the right person to comment on this, it was implemented by him. As per POSIX requirement all types should be inside types.h file. So I mode it from pthread.h --> types.h file

andyross replied:
It's just a typesafety thing.  If you typedef it then it becomes legal to pass a semaphore to pthread_mutex_unlock(), which doesn't work in general.  You can implement a mutex in terms of a semaphore, but not the reverse.

andyross replied:
Oh wait, is the question why it's a pointer and not a simple struct inclusion?  It has to do with our dlist declaration semantics: a zero-filled k_sem object is not a legal object.  You have to declare it with e.g. K_SEM_DEFINE.  So PTHREAD_MUTEX_DEFINE either needs to duplicate that code or use the existing macro, which requires that it store a pointer.

I had a patch to allow dlist to interpret NULL fields as "empty" (not zero cost, sadly) which would fix that.  Not sure if it's worth revisiting, or alternatively just fix pthread_mutex_init() to initialize the semaphore/dlist fields directly.  We should file a bug.

andyross replied:
Uh... ignore what I wrote.  We need to fix this now.

As written (yeah, by me), pthread_mutex_init() is going to call k_sem_init() on a garbage pointer on any mutex object that was not created with PTHREAD_MUTEX_DEFINE (which is itself a zephyr-specific API).   That's a pretty bad bug.  We should make the field an included struct, not a pointer, and change PTHREAD_MUTEX_DEFINE so that it initializes the dlist fields the same way K_SEM_DEFINE does.

andrewboie commented on include/posix/sys/types.h:
```diff
@@ -0,0 +1,74 @@
+	struct k_sem *sem;
+} pthread_mutex_t;
+
+typedef struct pthread_mutexattr {
```
so make this an empty struct instead of wasting 4 bytes on "unused"

youvedeep replied:
@andyross  implementation. We can take this as a fix later.

andrewboie commented on include/posix/sys/types.h:
```diff
@@ -0,0 +1,74 @@
+} pthread_cond_t;
+
+typedef struct pthread_condattr {
+	int unused;
```
so make this an empty struct instead of wasting 4 bytes on "unused"

youvedeep replied:
@andyross implementation. We can take this as a fix later.

andyross replied:
I had some minor reason (I'm guessing future binary compatibility with something that wanted a flags word?), but it's not important.  The "unused" fields are all part of attribute structs, not the runtime ones.  In general, these are always either passed as a null pointer or they live on the stack of initialization code which isn't likely to overflow.

Obviosuly none of the existing functions use the attribute structs at all, so it hurts nothing to remove that field.  No personal preference.

andrewboie replied:
@youvedeep please just delete the 'unused' members. thanks

youvedeep replied:
Sure @andrewboie I will remove unused members.

andrewboie commented on include/posix/sys/types.h:
```diff
@@ -0,0 +1,74 @@
+} pthread_barrier_t;
+
+typedef struct pthread_barrierattr {
+	int unused;
```
so make this an empty struct instead of wasting 4 bytes on "unused"

youvedeep replied:
@andyross implementation. We can take this as a fix later.

youvedeep-singh committed:
kernel: POSIX: Fixing return value of POSIX APIs on error.

As per IEEE 1003.1 POSIX APIs should return ERROR_CODE on error.
But currently these are returning -ERROR_CODE instead of ERROR_CODE.
So fixing the return value.

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

youvedeep commented:
Fixed the return ERROR_CODE  on error occurrence.

youvedeep commented:
There was a bug around pthread_cancel, So fixed in latest patch set.

youvedeep commented:
Please let me know if there are further comments on this patch set.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,537 @@
+	/* FIXME: Thread space allocation need to use mempool for better
+	 * space utilization under limited number of threads.
+	 */
+	if (thread_index < CONFIG_MAX_PTHREAD_COUNT) {
```
I would really want to see mempool being used for pthreads: it shouldn't be hard to use it (and if it is, we need to know why and fix it), and we'll be able to recycle handles when a thread ends.

youvedeep replied:
Sure I will change it to mempool. I wanted to take this as enhancement but based on comments I will make this change in new version of patch today.

lpereira replied:
Creating a thread is not something that's latency-sensitive.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,537 @@
+			thread->state = PTHREAD_TERMINATED;
+			pthread_mutex_unlock(&thread->cancel_lock);
+		} else {
+			thread->retval = PTHREAD_CANCELED;
```
Isn't PTHREAD_CANCELED a state?

youvedeep replied:
PTHREAD_CANCEL_ENABLE is a state and PTHREAD_CANCELED is a return value.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,537 @@
+	cur_prio = k_thread_priority_get(*thread);
+
+	if (is_posix_prio_valid(new_prio, policy) == false ||
+	    new_prio == cur_prio) {
```
This function should not fail if the new priority is the same as the old priority. Either return 0 in this case, or remove the test and let k_thread_priority_set() do its thing.

youvedeep replied:
Yes understand it may confuse POSIX developer if she is trying to set the same value again,
I will make the change.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,537 @@
+int pthread_attr_init(pthread_attr_t *attr)
+{
+
+	if (attr->initialized == true) {
```
Since the results of calling this function with an already-initialized attr is undefined, let's keep it simple and remove this check.

youvedeep replied:
I observed at some urls that as per POSIX 1003.1 if we reinitialize an already initialized thread then it should  return EBUSY. please have a look at following :- 
http://pubs.opengroup.org/onlinepubs/007904975/functions/pthread_attr_destroy.html

lpereira replied:
The POSIX documentation you linked is the deprecated 2004 version. The current version is this one: http://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_attr_destroy.html 

In any case, your second point makes sense -- the newer version says that the results are undefined, so it's fine to return EBUSY in this case.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,537 @@
+ */
+int pthread_setcancelstate(int state, int *oldstate)
+{
+	struct posix_thread *pthread = (struct posix_thread *) pthread_self();
```
pthread_self() is currently just returning whatever k_current_get() returns. That's fine, but if the current thread isn't a pthread, the type punning is going to be undefined behavior and this might lead to memory corruption if it's called from a thread that's not been created with pthread_create(). I suggest implementing something like "_is_thread_pthread()", similar to "_is_in_isr()", and return an error before proceeding.

youvedeep replied:
hmm

But we are calling pthread_self() only from a POSIX. So it will always return a thread id which is corresponding to a POSIX thread.
 
Do you want to this implemented through creating a entry (something like pthread_valid into pthreadObject) or through some other way?

lpereira replied:
This function can still be called from a thread not created by pthread_create().

lpereira commented on include/posix/pthread.h:
```diff
@@ -7,38 +7,57 @@
-	_wait_q_t wait_q;
-} pthread_cond_t;
+/* Pthread cancellation */
+#define PTHREAD_CANCEL_POS	0
```
Any #define that is an implementation detail should not be exported. Move these to the .c file, or use another namespace for them (e.g. prefix with _).

youvedeep replied:
Sure I will modify this,

lpereira commented on include/posix/pthread.h:
```diff
@@ -193,7 +212,7 @@ static inline int pthread_mutex_timedlock(pthread_mutex_t *m,
 	int ret = k_sem_take(m->sem, _ts_to_ms(to));
 
-	return ret == -EAGAIN ? -ETIMEDOUT : ret;
+	return ret == EAGAIN ? ETIMEDOUT : ret;
```
k_sem_take() will return a negative error code. Error codes should be translated correctly here, you can't just return whatever it returns.

lpereira replied:
In fact, this is the same for other APIs here (I can't comment on pthread_mutex_lock(), etc).  I did comment on this a few times already, and even provided code in comments to help convert the error codes.

youvedeep replied:
k_sem_take is going to return 
    1. 0 --> sucess
    2. -EAGAIN or -EBUSY 
So return value of "pthread_mutex_timedlock" i have modified  to 
return ret != 0 ? ETIMEDOUT : ret;

andyross commented on tests/kernel/pthread/src/pthread.c:
```diff
@@ -26,18 +28,6 @@ PTHREAD_BARRIER_DEFINE(barrier, N_THR);
-
-K_THREAD_DEFINE(thread2, STACKSZ, thread_top, (void *)2, NULL, NULL,
-		K_HIGHEST_THREAD_PRIO, 0, K_NO_WAIT);
-
```
I don't have any particular complaints with any of these changes.  But it might be worth adding them as a separate test.  Using pthread synchronization with Zephyr-native threads is a supported use case, and this removes testing for that while it adds testing for POSIX-native thread APIs.  Why not both?

youvedeep replied:
Idle here was to convert the pthread application entirely POSIX compliant and avoid using Zephyr APIs.
Is your suggestion to keep this application untouched and add posix compliant version of this application (pthread) separately ?

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,537 @@
+#include "ksched.h"
+#include "wait_q.h"
+
+const pthread_attr_t init_pthread_attrs = {
```
This should be static. It's an implementation detail that's used only by functions in this .c file

youvedeep replied:
Yes it should be static.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,537 @@
+{
+	struct posix_thread *thread = (struct posix_thread *) pthread;
+
+	if (thread == NULL || thread->state == PTHREAD_TERMINATED) {
```
Yeah, we really need a way to determine if a pthread handle is indeed a pthread handle. It's possible to improve the mempool API to say if a given pointer is part of that pool and use this (it should be efficient, it's just a range check).

youvedeep replied:
@lpereira  Is there some API in zephyr to check if a pointer to memblock is a part of mem pool ?

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,537 @@
+	pthread_mutex_lock(&self->cancel_lock);
+	if (self->cancel_state == PTHREAD_CANCEL_DISABLE) {
+		self->cancel_state = PTHREAD_CANCEL_ENABLE;
+		_thread_essential_clear();
```
This function is called by the pthread wrapper, so it might be a good idea to mention that pthread won't work at all in user mode at the moment.

youvedeep replied:
I was thinking of removing this reference to  _thread_essential_clear and _thread_essential_set and handle through PTHREAD_CANCEL_ENABLE entirely.

Yes as of now the POSIX implementation will not work with user mode and there are other candidates too which are not user mode compliant. Example: mempool too is not user mode compliant.

andyross commented on include/posix/unistd.h:
```diff
@@ -0,0 +1,46 @@
+static inline int usleep(useconds_t useconds)
+{
+	/* Creating Busy wait for time below msecs */
+	k_busy_wait(useconds % USEC_PER_MSEC);
```
Remove the busy wait.  Precise timing at the expense of wasted cycles is not the intent of usleep() on any existing platforms I know of.

And if you insist on doing this for precision, then this is the wrong way to do it anyway:  the k_sleep() call will only be precise to within a timer interrupt, which is tick-granular on most systems (i.e. much coarser than a microssecond).  So what you have to do is k_sleep() *first* (probably with some slop subtracted so you don't overshoot) then check the elapsed time with k_cycle_get_32() and busy wait on that until the desired time has expired.

But you don't want to do that.  Just yank the busy wait.

youvedeep replied:
hmm I got your idea. 

My intention here was :- 
if time in usec is below 1 msec ==> Create with busy wait.
If it is above or equal to 1 msec ==> create delay with k_sleep

As of now my logic was taking care of first case completely (when usecs are below 1msecs) but it is getting unfair to case where it is greater than 1msec. I will mention this clearly.

I was thinking in another way if delay to give is more than 1 msec shall we take the celling and  create the delay using usec. Something like usecs = 2.5msec then create delay k_sleep(3). This will make sure that thread is not scheduled before time. 
What is your suggestion on this ?

lpereira commented on kernel/posix/pthread_cond.c:
```diff
@@ -33,7 +33,7 @@ static int cond_wait(pthread_cond_t *cv, pthread_mutex_t *mut, int timeout)
 	pthread_mutex_lock(mut);
 
-	return ret == -EAGAIN ? -ETIMEDOUT : ret;
+	return ret == EAGAIN ? ETIMEDOUT : ret;
```
The return code from _Swap() is a negative error code. Please revisit the error code conversion.

andyross commented on tests/kernel/pthread/src/pthread.c:
```diff
@@ -173,14 +173,23 @@ int barrier_test_done(void)
+	if (min_prio < 0 || max_prio < 0 || schedparam.priority < min_prio ||
+	    schedparam.priority > max_prio) {
+		TC_START("Scheduling priority outside valid priority range\n");
+		goto done;
```
This doesn't actually test the scheduler behavior, which is kind of implied in the commit message.  If we are going to support POSIX scheduler behavior, we should be writing a test to make sure that it does what the API claims.

youvedeep replied:
yes i got it, I should have give this comment through TC_PRINT.

andyross reviewed (changes requested):
Mostly whining.  One legitimate issue with the way sleep is implemented

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,537 @@
+		pthread_cond_wait(&pthread->state_cond, &pthread->state_lock);
+	}
+
+	if (pthread->state == PTHREAD_EXITED) {
```
Use an `else if` here.

youvedeep replied:
PTHREAD_JOINABLE and PTHREAD_EXITED are not implemented through ifelse as status of thread will change from PTHREAD_JOINABLE --> PTHREAD_EXITED  on pthread_cond_broadcast (inside pthread_exit). 
Example : 
1. Thread T1 calls pthread_join for thread T2.
2. Thread T1 waits at pthread_cond_wait as T2 has not exited.
.
.
3. Thread T2 calls pthread_exit and it updates self->state and retvalue. and calls pthread_cond_broadcast.
4. Now thread T1 receives the condition variable.
5. It enters into pthread->state == PTHREAD_EXITED and gets the return value,

So a thread need to enter into both the conditions, so it has not been put through else if.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,537 @@
+	pthread_mutex_lock(&pthread->state_lock);
+
+	if (pthread->state == PTHREAD_JOINABLE) {
+		pthread_cond_wait(&pthread->state_cond, &pthread->state_lock);
```
This should return 0 here

youvedeep replied:
when the state if thread will change from PTHREAD_JOINABLE to PTHREAD_EXITED. it will pass through pthread->state == PTHREAD_EXITED) and will eventually return 0.

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,537 @@
+ *
+ * See IEEE 1003.1
+ */
+int pthread_join(pthread_t thread, void **status)
```
For this function, it might be better to have a `int ret` that's set by each block comparing with `pthread->state`. This will reduce its size slightly, and have only one mutex_unlock/return block.

youvedeep replied:
sure i will modify this.

youvedeep-singh committed:
tests: kernel: posix: pthread: Add pthread test.

This test is POSIX based implementation of tests:kernel:pthread test.
It used POSIX APIs instead of Zephyr APIs.

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

youvedeep-singh committed:
kernel: POSIX: removing unused elements from POSIX object attribute.

This patch removes unused member element from POSIX object attributes
(mutex, condition variable and barrier).

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

youvedeep commented:
Addressed comments received (except @andrew's comment to fix mutex Object issue). Can you please review it further and provide feedbacks.

youvedeep-singh committed:
tests: kernel: posix: pthread_cancel: POSIX thread cancel test.

Added test for POSIX thread cancel APIs.

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

youvedeep commented:
1. implemented memory allocation for pthread object using mempool.
2. Added pthread_cancel test.

youvedeep commented:
Guys Can you please give further comments or provide feedback on this PR?

lpereira commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,531 @@
+};
+
+/* Memory pool for pthread space */
+K_MEM_POOL_DEFINE(posix_thread_pool, sizeof(struct posix_thread),
```
Nothing is returned to this memory pool; we'll still run out of pthread handles.

I remember threads having a callback function that got called whenever they were destroyed.  Can't we used that to return something to the pool?  It might be only for threads that are aborted, which is fine; pthread_exit() can schedule a call to k_mem_pool_free() (it can't call it directly).

youvedeep replied:
Currently I am not returning back to mem pool becuase we need to keep pthread object to handle 
pthread_join requests. If we will free the space then pthread_join will have issue.

But I agree we need to figure out a way to return back to mem pool on pthread termination. I can explore it further. But even when k_thread is destroyed, we may need to access pthread object such as retval, pthtread state, statue_mutex to address pthread_join.

lpereira reviewed (commented):
Looking a lot better, thanks!

youvedeep commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,531 @@
+	prio = posix_to_zephyr_priority(attr->priority, attr->schedpolicy);
+
+	if (k_mem_pool_alloc(&posix_thread_pool, &block,
+			     sizeof(sizeof(struct posix_thread)), 100) == 0) {
```
this is a typo, thanks for catching it. I will fix this.

youvedeep commented on kernel/posix/pthread.c:
```diff
@@ -0,0 +1,531 @@
+	struct posix_thread *thread;
+	struct k_mem_block block;
+
+	/*
```
sure I put comment in doxygen.

youvedeep-singh committed:
kernel: POSIX: Compatibility layer for POSIX timer APIs.

This patch provides POSIX timer APIs for POSIX 1003.1 PSE52 standard.

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

youvedeep-singh committed:
tests: kernel: posix: timer: POSIX timer test.

Added test for POSIX timer APIs.

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

youvedeep commented:
Addressed comments from @lpereira .
Also Added 2 patches for : 
POSIX Timer APIs (12/13)
POSIX timer test API (13/13)

youvedeep-singh committed:
arch: POSIX: Add guard into posix_cheats file.

Some of APIs of POSIX implmentation layer has same name as
native_posix architecture. posix_cheats.h is used to handle this
duplication in API name. Adding a guard in posix_cheats.h based on
CONFIG_PTHREAD_API.

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>
